src/components/Table/hoc/withTableSettings/TableColumnSetup/TableColumnSetup.tsx

```diff
@@ -1,270 +1,284 @@
-        (list: Item[]) =>
-        ({oldIndex, newIndex}: {oldIndex: number; newIndex: number}) => {
-            setCurrentItems(List.moveListElement(list.slice(), oldIndex, newIndex));
-        };
+    const dndRenderContainer = useDndRenderContainer({onApply, onDragEnd});
 
-    const handleUpdate = (value: Item[]) => setCurrentItems(value);
+    const dndRenderItem = useDndRenderItem(sortable);
 
-    const handleClosePopup = () => setInitialState();
+    const renderControl: TreeSelectProps<unknown>['renderControl'] = ({toggleOpen}) => {
+        const onKeyDown = createOnKeyDownHandler(toggleOpen);
 
-    const handleControlClick = React.useCallback(() => {
-        if (!disabled) {
-            setFocused(!focused);
-            setRequiredItems(getRequiredItems(items));
-            setCurrentItems(getConfigurableItems(items));
-        }
-    }, [disabled, focused, items]);
-
-    const handleApplyClick = () => {
-        setInitialState();
-
-        const newItems = requiredItems.concat(currentItems);
-
-        if (items !== newItems) {
-            props.onUpdate(newItems);
-        }
-    };
-
-    const handleItemClick = (value: Item) => {
-        const newItems = currentItems.map((item) =>
-            item === value ? {...item, selected: !item.selected} : item,
-        );
-        handleUpdate(newItems);
-    };
-
-    const renderItem = (item: Item) => {
         return (
-            <div className={b('item-content')}>
-                {item.required ? (
-                    <div className={b('lock-wrap', {visible: item.selected})}>
-                        <Icon data={Lock} />
-                    </div>
-                ) : (
-                    <div className={b('tick-wrap', {visible: item.selected})}>
-                        <Icon data={Check} className={b('tick')} width={10} height={10} />
-                    </div>
-                )}
-                <div className={b('title')}>{getItemTitle(item)}</div>
-            </div>
+            renderSwitcher?.({onClick: toggleOpen, onKeyDown}) || (
+                <Button onClick={toggleOpen} extraProps={{onKeyDown}}>
+                    <Icon data={Gear} />
+                    {i18n('button_switcher')}
+                </Button>
+            )
         );
     };
 
-    const renderStatus = () => {
-        if (!showStatus) {
-            return null;
-        }
-
-        const selected = getCountSelected();
-        const all = propsItems.length;
-        const status = `${selected}/${all}`;
-
-        return <span className={b('status')}>{status}</span>;
-    };
-
-    const renderRequiredColumns = () => {
-        const hasRequiredColumns = requiredItems.length;
+    const onOpenChange = (open: boolean) => {
+        setOpen(open);
 
-        if (!hasRequiredColumns) {
-            return null;
+        if (open === false) {
+            // we want to set items to initial and anyone who subscribed on items know it
+            const newItems = [...propsItems];
```

Comment:
it is possible to avoid this, but in that cause `valueRef` and `dndItemsRef` should be wrapped on `useMemo(() => {}, [items])`.

In my taste this way is better

Reply:
It's probably better to `useMemo` it here.

3240a14161ea5adbf68286c7e18f01825e52f661